Have the Fakeclient implement the resources interfaces

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
VM, VMI, VMIM, KV and VMIRS fakeclients do not implement VM, VMI, VMIM, KV and VMIRS interfaces, resulting in the impossibility of using it for unit tests.

After this PR:
VM, VMI, VMIM, KV and VMIRS fakeclients implement VM, VMI, VMIM, KV and VMIRS interfaces so that we can start using it in unit tests.
Also, the go version of the client-go has been updated to 1.21

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Updated go version of the client-go to 1.21
```

xpivarc commented on hack/generate.sh:
```diff
@@ -7,7 +7,7 @@ source $(dirname "$0")/config.sh
 # generate clients
 CLIENT_GEN_BASE=kubevirt.io/client-go/generated
-rm -rf ${KUBEVIRT_DIR}/staging/src/${CLIENT_GEN_BASE}
+#rm -rf ${KUBEVIRT_DIR}/staging/src/${CLIENT_GEN_BASE}
```
Remove, rather than comment. Do we know what will be re-generated and what not?

xpivarc commented on staging/src/kubevirt.io/client-go/generated/kubevirt/clientset/versioned/typed/core/v1/fake/fake_virtualmachine_expansion.go:
```diff
@@ -0,0 +1,120 @@
+
+func (c *FakeVirtualMachines) Restart(ctx context.Context, name string, restartOptions *v1.RestartOptions) error {
+	_, err := c.Fake.
+		Invokes(fake2.NewPutSubresourceAction(virtualmachinesResource, c.ns, "restart", name), nil)
```
I think it would be great if we can have a new struct, let's call it `restartActionImpl` that would also provide the options. Then a reactor could access these as well.

xpivarc reviewed (commented):
Few comments but I would approve on this approach

acardace commented:
/lgtm

this looks great! Nice use of generics!

fossedihelm commented:
@acardace I split the putaction creation in a separate commit and added a new missing method into it: `NewRootPutAction`.
Also, I've added the interface:
```
type PutAction[T any] interface {
	testing.Action
	GetName() string
	GetOptions() T
}
```
and a method to retrieve the options:
```
func (a PutActionImpl[T]) GetOptions() T {
	return a.Options
}
```
This is to be aligned with other Actions in `k8s.io/client-go@v0.26.4/testing/actions.go`
PTAL

fossedihelm commented:
/hold
missing some steps

fossedihelm commented on staging/src/kubevirt.io/client-go/generated/kubevirt/clientset/versioned/typed/core/v1/virtualmachine_expansion.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func (c *virtualMachines) Restart(ctx context.Context, name string, restartOptions *v1.RestartOptions) error {
+	// TODO not implemented yet
```
I plan to resolve (most of) these TODOs in a follow-up PR.
This is to reduce the scope of the PR. WDYT?

xpivarc replied:
Maybe we should panic?

fossedihelm replied:
As per the offline discussion, let's address this as a follow-up PR, moving the implementation from VM file to here. Thanks

fossedihelm commented:
/unhold

fossedihelm commented:
Changes: rebase and addressed kv interface too

fossedihelm commented:
Changes: rebase to pick the operator-observability fix

xpivarc commented on staging/src/kubevirt.io/client-go/generated/kubevirt/clientset/versioned/typed/core/v1/streamer.go:
```diff
@@ -1,4 +1,23 @@
-package kubecli
+/*
+ * This file is part of the KubeVirt project
```
I think the streamer and websocket related logic could be separate package but I am not holding on this.

fossedihelm replied:
As per the offline discussion, let's address this as a follow-up PR. Thank you!

xpivarc reviewed (commented):
/approve

fossedihelm commented:
Changes: Added a commit to cover vmirs interface

acardace commented:
/lgtm
